tests/cli/test_output_option.py

```diff
@@ -17,97 +17,210 @@
 
     See https://github.com/SwissDataScienceCenter/renku-python/issues/387
     """
-    base_sh = ['sh', '-c', 'mkdir -p "$0"; touch "$0/$1"']
+    A_SCRIPT = ('sh', '-c', 'mkdir -p "$0"; touch "$0/$1"')
+    exit_code, cwl = cli('run', *A_SCRIPT, 'outdir', 'foo')
+
+    assert exit_code == 0
+    assert len(cwl.outputs) == 1
+    assert cwl.outputs[0].type == 'Directory'
+
+
+def test_explicit_inputs_must_exist(cli):
+    """Test explicit inputs exist before run"""
+    exit_code, _ = cli('run', '--input', 'foo', 'touch', 'bar')
+
+    assert exit_code == 1
+
+
+def test_explicit_inputs_are_inside_repo(cli):
+    """Test explicit inputs are inside the Renku repo"""
+    exit_code, _ = cli('run', '--input', '/tmp', 'touch', 'foo')
 
-    assert 0 == run(args=['run'] + base_sh + ['output', 'foo'])
-    assert (client.path / 'output' / 'foo').exists()
+    assert exit_code == 1
 
-    # There should be only one command line tool.
-    tools = list(client.workflow_path.glob('*_sh.cwl'))
-    assert 1 == len(tools)
 
-    with tools[0].open('r') as f:
-        cwl = CWLClass.from_cwl(yaml.safe_load(f))
+def test_explicit_inputs_and_outputs_are_listed(cli, client):
+    """Test explicit inputs and outputs will be in generated CWL file"""
+    cli('run', 'mkdir', 'foo')
+    cli('run', 'touch', 'bar', 'baz')
 
-    assert 1 == len(cwl.outputs)
-    assert 'Directory' == cwl.outputs[0].type
+    exit_code, cwl = cli(
+        'run', '--input', 'foo', '--input', 'bar', '--output', 'baz', 'echo'
+    )
+
+    assert exit_code == 0
+    assert len(cwl.inputs) == 2
+    cwl.inputs.sort(key=lambda e: e.type)
+    assert str(cwl.inputs[0].default) == '../../foo'
+    assert cwl.inputs[0].type == 'Directory'
+    assert cwl.inputs[0].inputBinding is None
+    assert str(cwl.inputs[1].default) == '../../bar'
+    assert cwl.inputs[1].type == 'File'
+    assert cwl.inputs[1].inputBinding is None
+    assert len(cwl.outputs) == 1
+    assert cwl.outputs[0].outputBinding.glob == 'baz'
+
+
+def test_explicit_inputs_can_be_in_inputs(cli):
+    """Test explicit inputs that are in inputs are treated as normal inputs"""
+    cli('run', 'touch', 'foo')
+
+    exit_code, cwl = cli('run', '--input', 'foo', '--no-output', 'ls', 'foo')
+
+    assert exit_code == 0
+    assert len(cwl.inputs) == 1
+    assert str(cwl.inputs[0].default) == '../../foo'
+    assert cwl.inputs[0].type == 'File'
+    assert cwl.inputs[0].inputBinding is not None
+
+
+def test_explicit_inputs_in_subdirectories(cli, client):
+    """Test explicit inputs that are in sub-dirs are made accessible"""
+
+    # Set up a script with hard dependency
+    cli('run', '--no-output', 'mkdir', 'foo')
+    exit_code, cwl = cli('run', 'sh', '-c', 'echo "some changes" > foo/bar')
+    exit_code, cwl = cli('run', 'sh', '-c', 'echo "cat foo/bar" > script.sh')
+
+    exit_code, cwl = cli(
+        'run', '--input', 'foo/bar', '--input', 'script.sh', 'sh', '-c',
```

Comment:
This should also work without specifying the scripy.py as an input I believe

Reply:
Actually this won't work without specifying the input. I dug the problem and the generated CWL looks like this:
```
baseCommand:
- sh
class: CommandLineTool
cwlVersion: v1.0
hints: []
inputs:
  input_1:
    default: sh script.sh > output
    inputBinding:
      position: 1
      prefix: -c
      separate: true
      shellQuote: true
    streamable: false
    type: string
...
```
As you can see `input_1` is not a File but a string which means that it won't be added to the explicit input automatically (i.e. by Andreas' commit). This is how the CWL is generated even before my PR (on master it generates the same).